test/infrastructure/test-extension/topologymutation/handler.go

```diff
@@ -0,0 +1,408 @@
+		return err
+	}
+
+	vsphereCluster.Spec.Template.Spec.ControlPlaneEndpoint.Host = controlPlaneIPAddr
+	vsphereCluster.Spec.Template.Spec.ControlPlaneEndpoint.Port = controlPlanePort
+
+	return nil
+}
+
+// patchGovmomiMachineTemplate patches the govmomi VSphereMachineTemplate.
+// NOTE: this patch is not required for any special reason, it is used for testing the patch machinery itself.
+func patchGovmomiMachineTemplate(_ context.Context, vsphereMachineTemplate *infrav1.VSphereMachineTemplate, templateVariables map[string]apiextensionsv1.JSON, isControlPlane bool) error {
+	// patch vSphereTemplate
+
+	var err error
+	vsphereMachineTemplate.Spec.Template.Spec.Template, err = calculateImageName(templateVariables, isControlPlane)
+
+	return err
+}
+
+// patchSupervisorMachineTemplate patches the supervisor VSphereMachineTemplate.
+// NOTE: this patch is not required for any special reason, it is used for testing the patch machinery itself.
+func patchSupervisorMachineTemplate(_ context.Context, vsphereMachineTemplate *vmwarev1.VSphereMachineTemplate, templateVariables map[string]apiextensionsv1.JSON, isControlPlane bool) error {
+	// patch vSphereTemplate
+
+	var err error
+	vsphereMachineTemplate.Spec.Template.Spec.ImageName, err = calculateImageName(templateVariables, isControlPlane)
+
+	return err
+}
+
+func calculateImageName(templateVariables map[string]apiextensionsv1.JSON, isControlPlane bool) (string, error) {
+	// patch vSphereTemplate
+	versionVariable := "builtin.controlPlane.version"
+	if !isControlPlane {
+		versionVariable = "builtin.machineDeployment.version"
+	}
+
+	version, err := topologymutation.GetStringVariable(templateVariables, versionVariable)
+	if err != nil {
+		return "", err
+	}
+
+	// Fallback to the v1.30.0 image, except for v1.28.0 and v1.29.0.
+	if version != "v1.28.0" && version != "v1.29.0" {
+		version = "v1.30.0"
+	}
+
+	return fmt.Sprintf("ubuntu-2204-kube-%s", version), nil
+}
+
+// ValidateTopology implements the HandlerFunc for the ValidateTopology hook.
+// Cluster API E2E currently are just validating the hook gets called.
+// NOTE: custom RuntimeExtension must implement the body of this func according to the specific use case.
+func (h *ExtensionHandlers) ValidateTopology(ctx context.Context, _ *runtimehooksv1.ValidateTopologyRequest, resp *runtimehooksv1.ValidateTopologyResponse) {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("ValidateTopology called")
+
+	resp.Status = runtimehooksv1.ResponseStatusSuccess
+}
+
+// DiscoverVariables implements the HandlerFunc for the DiscoverVariables hook.
+func (h *ExtensionHandlers) DiscoverVariables(ctx context.Context, _ *runtimehooksv1.DiscoverVariablesRequest, resp *runtimehooksv1.DiscoverVariablesResponse) {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("DiscoverVariables called")
+
+	resp.Status = runtimehooksv1.ResponseStatusSuccess
+
+	// The variables are a copy of what flavorgen generates in `packaging/flavorgen/flavors/clusterclass_generators.go`
```

Comment:
Was the plan to use GetClusterClassVariables here?

Reply:
I did copy over because:
* `packaging` is left out via Dockerignore (but we could move that function to e.g. `internal/clusterclass`.
* The function is written to have different outputs for supervisor/govmomi (but we could pass `false` which gets us all variables)

Happy to do the move though if we want :-)

Reply:
I'm fine with copying, but wondering if we should keep the if/else

Reply:
Should / could we un-export GetClusterClassVariables then again?

Reply:
done. Its now set at the clusterclass which mode to use and also the GetClusterClassVariables is now used from the internal location.